### ISSUE TITLE
[Question] Intercept DS4 input and return a new input?

### ISSUE THREAD

**Issue by RobertHodan** [Ryochan7/DS4Windows#446]
My goal is to find where the inputs are handled, pass the controller inputs into a method, return new inputs, and then have DS4Windows continue as it normally would have.

The tricky thing is, pressing Square may need to return D-pad Up + Circle. Or pressing a combination of R2 + Square may need to return another combination of inputs. Pressing R1 may very well act like a toggle, so that R1 remains active until the next time R1 is pressed again.

Long story short, I want to be able to do a lot of silly things within that method. I would greatly appreciate it if anyone can point me in the right direction, or give some explanations as to what files / classes I want to look into.

**Comment by Ryochan7** [comment]
Unfortunately, there is no support for multiple button outputs for one
input. The macro system offers very limited functionality for multiple
outputs but it is not very useful.

**Comment by RobertHodan** [comment]
Thanks for the reply, that's unfortunate to hear.

I'm not familiar with C#, and my experience with any C language is limited... but I've been looking through the "performDs4Input" method, and I'm hoping it'll be as simple as passing cState into method, and returning it with the new values.

It's not clean by any means, but I just want to have some fun with it. I'll be closing this since I've found what I was looking for.

**Comment by Ryochan7** [comment]
The main mapping is done in the Mapping.MapCustom method in DS4Windows\DS4Control\Mapping.cs. Despite changes to help improve it, it is still a mess. Also, the DS4ControlSettings class is currently set up to only support mapping to a single main action and a possible shift action. Changing that class probably wouldn't be so bad but I avoid editing the MapCustom method as much as possible.

**Comment by RobertHodan** [comment]
Thanks for the suggestions.  I'm only interested in remapping the buttons on my DS4 controller, and modifying Mapping.MapCustom seems like a clean solution considering my goal.

To give you some context of what I ultimately want to do... I just want to make a simple macro / button remap class for a game called "Devil May Cry 4", in preparation for the upcoming "Devil May Cry 5". Instead of using the D-Pad for changing styles, this will be performed by specific button inputs. By default (no buttons held down), the "Swordsmaster" style will be equipped. If L2 is pressed, "Trickster" would be selected, and pressing 'X' during this would output 'Circle'. If 'Square' is pressed, then "Gunslinger" would be equipped, and 'Circle' would be output instead. Long story short, I just want to simplify the control scheme for myself.

As far as modifying the source code, I don't intend to make any real changes to it. I just want to pass "cState" into an object, and mutate it to get the desired behaviour. At the moment, it's just two additional lines like so:
`
private static DMCRemap dmcRemap = new DMCRemap();

public static void MapCustom(int device, DS4State cState, DS4State MappedState, DS4StateExposed eState, Mouse tp, ControlService ctrl)
{
        dmcRemap.update(cState);
...
`

I think it would be pretty amazing to have some kind of script support for cState mutations, and who knows... if a feature like that would be within the scope of this repository, I think it would be a fun feature to tackle. As it stands, I want to avoid any needless work, and keep my changes to just those two lines above. Simply instantiating my class, and calling its update method.
